ocaml/xcp-rrdd/bin/rrdp-netdev/rrdp_netdev.ml

```diff
@@ -132,18 +135,16 @@ let transform_taps devs =
     newdevnames
 
 let generate_netdev_dss () =
-  let _, doms, _ =
-    Xenctrl.with_intf (fun xc -> Xenctrl_lib.domain_snapshot xc)
-  in
-
-  let uuid_of_domid domains domid =
-    let _, uuid, _ =
-      try List.find (fun (_, _, domid') -> domid = domid') domains
-      with Not_found ->
-        failwith
-          (Printf.sprintf "Failed to find uuid corresponding to domid: %d" domid)
-    in
-    uuid
+  let uuid_of_domid domid =
+    try
+      Xenstore.with_xs (fun xs ->
+          let vm = xs.Xenstore.Xs.getdomainpath domid ^ "/vm" in
+          let vm_dir = xs.Xenstore.Xs.read vm in
+          xs.Xenstore.Xs.read (vm_dir ^ "/uuid")
```

Comment:
I know I'm late to the party here, but this is very expensive.  You've got 3 xenstore accesses to do what was previously a single hypercall (would be faster than a single xenstore access), and xenstore is a known bottleneck.

First, `getdomainpath $x` is universally `/local/domain/$x`.  Some original plans for xenstore involved having `/remote/$host/domain/$domid` as shared-virtual-memory across the pool.  Luckily, they never even got started, and you can reasonably hard-code it.

Second, `/vm/$uuid/uuid` is always self-referential, so you could just split the answer from `/local/domain/$domid/vm` at `/` and take the final component.

Or, if you want something less dodgy, just have Xapi/Xenops/whatever fill in a new top-level dir with `/uuid-by-domid/$domid = $uuid` so logic like this can issue a single read for `/uuid-by-domid/$domid` and get the precise data you want.